packages/flutter_tools/lib/src/daemon.dart

```diff
@@ -0,0 +1,242 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'base/common.dart';
+import 'base/io.dart';
+import 'base/logger.dart';
+import 'base/utils.dart';
+import 'convert.dart';
+import 'globals.dart' as globals;
+
+/// Parse binary streams in the JSON RPC format understood by the daemon, and
+/// convert it into a stream of JSON RPC messages.
+Stream<Map<String, Object?>> _convertInputStream(Stream<List<int>> inputStream) {
+  return utf8.decoder.bind(inputStream)
+  .transform<String>(const LineSplitter())
+  .where((String line) => line.startsWith('[{') && line.endsWith('}]'))
+  .map<Map<String, Object?>?>((String line) {
+    line = line.substring(1, line.length - 1);
+    return castStringKeyedMap(json.decode(line));
+  })
+  .where((Map<String, Object?>? entry) => entry != null)
+  .cast<Map<String, Object?>>();
+}
+
+/// A stream that a [DaemonConnection] uses to communicate with each other.
+abstract class DaemonStreams {
+  /// Stream that contains input to the [DaemonConnection].
+  Stream<Map<String, Object?>> get inputStream;
+
+  /// Outputs a message through the connection.
+  void send(Map<String, Object?> message);
+
+  /// Cleans up any resources used.
+  Future<void> dispose() async { }
+}
+
+/// A [DaemonStream] that uses stdin and stdout as the underlying streams.
+class StdioDaemonStreams extends DaemonStreams {
+  StdioDaemonStreams([Stdio? stdio]) :
+    _stdio = stdio ?? globals.stdio,
+    inputStream = _convertInputStream((stdio ?? globals.stdio).stdin);
```

Comment:
Make the injected globals required, ideally we should be able to remove `import 'globals.dart' as globals;` from almost everywhere but `context_runner.dart`.  Unfortunately the commands don't have their dependencies injected yet, so the global context objects will need to be passed from there.
```suggestion
  StdioDaemonStreams(Stdio stdio) :
    _stdio = stdio,
    inputStream = _convertInputStream(stdio.stdin);
```